### PR TITLE
java-pkg-simple.eclass: support multi-release according to JEP 238

### PR DESCRIPTION
**Pull request by vaukai**
All new functionality is hidden behind a condition `if [[ -n ${JAVA_RELEASE_SRC_DIRS[@]}` which should ensure that existing ebulds cannot run into it. There is still room for improvements (javadocs, resources, tests, ...) but for the `ebuilds` which are attached in the further commits it works without problems.

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

**Comment by vaukai** [comment]
Didn't know about the `git show --ignore-space-change` option. Will just fix the indentation now.

**Comment by vaukai** [comment]
@fordfrog for viewing the `eclass` clanges it needs to [hide whitespace](https://github.com/gentoo/gentoo/pull/39805/commits/af26e66a3fce9ab84376d37f2bd3de4fad1bf51b?diff=unified&w=1).
and, this is now my personal prio 1.

**Comment by vaukai** [comment]
Dropped changing `commons-codec`. Will bump to `commons-codec-1.17.2` once the eclass change is in the tree.

**Comment by fordfrog** [comment]
i just went through the eclass changes. i think we should also have there a fast-fail test that checks whether minimum java version matches the max java version used in the multi-jar specificatoin. i guess it would be better for developers to get a clear message what to do (use higher java version for compilation). once we finish this, i'll start to do some testing.

**Comment by fordfrog** [comment]
> Guess I need to get the max Java from the `associative array` and compare with the active vm?

yes, exactly.

**Comment by fordfrog** [comment]
well, when thinking about it, it would be better to check the max version from the array to match the min DEPEND version from the ebuild. as if i have jdk 21 set as the default but i'd specify min java in DEPEND as 1.8, it would not trigger as version 21 would be used at runtime.

**Comment by vaukai** [comment]
> well, when thinking about it, it would be better to check the max version from the array to match the min DEPEND version from the ebuild. as if i have jdk 21 set as the default but i'd specify min java in DEPEND as 1.8, it would not trigger as version 21 would be used at runtime.


We already have one line 
```
		if ! java-pkg_is-vm-version-ge "9" ; then
```
which I think does what it should.
So if we used a similar mechanism a little further down, like
```
	# compile content of release-specific source directories
	local version
	for version in "${!JAVA_RELEASE_SRC_DIRS[@]}"; do
		local release="${version}"
		if ! java-pkg_is-vm-version-ge "${release}" ; then	
				die || "Wrong vm version, blah blah blah"
		fi
```
Would this work? I am presently not clear about how to get `the max version from the array`.

**Comment by vaukai** [comment]
> i think this would not work as intended as `java-pkg_is-vm-version-ge` returns the current vm, not the minimum java version required by the ebuild. so as i wrote, if DEPEND contains version 1.8 and i have version 21 as the default, version 21 is gonna be picked up instead of the lowest possible and the loop would not reveal the issue in the ebuild. so the loop might work, but we need to use min java version as defined in DEPEND rather than currently chosen java version.

Are we talking about different topics?
My understanding is that for compiling e.g https://github.com/bcgit/bc-java/tree/r1rv79/prov/src/main/jdk21 we need to ensure building with 21, means having at least jdk-21 installed.

**Comment by fordfrog** [comment]
> Are we talking about different topics? My understanding is that for compiling e.g https://github.com/bcgit/bc-java/tree/r1rv79/prov/src/main/jdk21 we need to ensure building with 21, means having at least jdk-21 installed.

generally, if the package needs to use java 21, we must ensure not only that the jdk used for compilation is at least 21, but we should also ensure that the ebuild does not specify lower version in DEPEND, as that is wrong. DEPEND must also have version 21.

**Comment by vaukai** [comment]
> > Are we talking about different topics? My understanding is that for compiling e.g https://github.com/bcgit/bc-java/tree/r1rv79/prov/src/main/jdk21 we need to ensure building with 21, means having at least jdk-21 installed.
> 
> generally, if the package needs to use java 21, we must ensure not only that the jdk used for compilation is at least 21, but we should also ensure that the ebuild does not specify lower version in DEPEND, as that is wrong. DEPEND must also have version 21.

Got it and did it as a separate commit.

File: eclass/java-pkg-simple.eclass [hunk 1]
```diff
@@ -543,6 +543,22 @@ java-pkg-simple_src_compile() {
...
+		for key in "${!JAVA_RELEASE_SRC_DIRS[@]}"; do
+		    # Compare the numeric value of the key
+		    if [[ key > highest_key ]]; then
+		        highest_key="$key"
+		    fi
+		done
+
+		local jdk="$(depend-java-query --get-lowest "${DEPEND}")"
+		if [[ "${jdk#1.}" -lt "${highest_key}" ]]; then
+			die || "Wrong DEPEND, needs higher version in >=virtual/jdk"
```
Comment (fordfrog) [inline eclass/java-pkg-simple.eclass:559]: nice! how about putting `${highest_key}` in the error message? also there's the redundant `||` for the `die` statements. it should be just `die <message>`.

**Comment by vaukai** [comment]
Draft mode again. Needs the logic changed to have a better trigger for `generate-module-info`.

**Comment by vaukai** [comment]
Closing this PR until the pieces are complete.
